llvm/lib/Target/AMDGPU/GCNHazardRecognizer.cpp

```diff
@@ -325,7 +330,9 @@ unsigned GCNHazardRecognizer::PreEmitNoops(MachineInstr *MI) {
   unsigned W = PreEmitNoopsCommon(MI);
   fixHazards(MI);
   CurrCycleInstr = nullptr;
-  return W;
+  unsigned NopPad =
+      NopPadding.getNumOccurrences() && !MI->isTerminator() ? NopPadding : 0;
```

Comment:
Why don't you want nop before a terminator?

Reply:
There can't be that between terminators.

Reply:
Because of a rule that all terminators have to be consecutive at the end of the block?

You could still have a nop before the first terminator.

Reply:
For the debug purpose, I'd say that would be fine because if hazard happens, the first instruction in the following BB has a nop before it.